cilium: Reinitialize datapath before annotating node object

jrfastab opened the pull request:
When rolling cilium-agent or doing an upgrade while running stress test
with encryption a small number of NoStateIn errors are seen. To capture
the error state (a cilium_host IP without an xfrm state rule) you need
to get into the pod near pod init and get somewhat lucky that init
took some longer time. For example I ran `ip x s` in a pod about
15seconds after launch and captured a case with new XfrmInNoErrors,
a cilium_host ip assigned, but no xfrm state rule for it. The packets
received are dropped.

The conclusion is remote nodes learn the new router IP before we have
the xfrm state rule loaded. The remote nodes then start using that
IP for the IPSec tunnel outer IP resulting in the errors when they
reach the local node without the xfrm rule yet. The errors eventually
resolve, but some packets are lost in the meantime.

The reason this happens is because first we configure the datapath
after we push node object updates. This is wrong because we need
to init the ipsec code path before we teach remote nodes about the
new IP. And second the configuration of the datapath does a lookup
in the node objects IPAddresses{} this is only populated from the
k8s watcher in the tunnel case. So we only have the fully populated
node object after we receive it through the k8s watcher. Again its
possible other nodes already have seen the event and started pushing
traffic with the new IPs.

To resolve push IPSec init code to create xfrm rules needed with
the new IPs before we publish them to the k8s node object. And
instead of pulling the IPs out of the node object simply pull them
directly from the node module. This resolves the XfrmInNoState and
XfrmIn*Policy* errors I've seen.

To reproduce the errors I can consistently reproduce with about
30 nodes, with httpperf test running from a pod in all nodes, and
then doing a 'rollout' of the cilium agent for awhile. Seems
a 2-3 hours almost ensures errors pop up. Usually the errors
happen much sooner. Initially I saw these errors on upgrade tests
which is another method to reproduce.

zacharysarah commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -174,7 +174,8 @@ func (n *NodeDiscovery) StartDiscovery() {
 	n.fillLocalNode()
-
+	n.Manager.NodeUpdated(n.localNode)
+	// Ensure local nodeUpdates are completed before we add local node to cluster.
```
Optional docs improvement
```suggestion
	// Ensure local nodeUpdates are completed before adding the local node to the cluster.
```

zacharysarah reviewed (dismissed):
Optional fix, otherwise LGTM for docs

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -941,6 +941,14 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+	// allocateIPs got us the routerIP so now we can create ipsec endpoint
+	// we must do this before publishing the router IP otherwise remote
+	// nodes could pick up the IP and send us outer headers we do not yet
+	// have xfrm rules for.
```
nit: I'd probably add a comment to the effect of this to ensure later refactors can take into account the dependencies for this new logic:
```suggestion
	// have xfrm rules for.
	//
	// This must occur strictly after `setupIPSec()` and `d.allocateIPs()`.
```

jrfastab replied:
will do.

joestringer commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -1081,3 +1081,16 @@ func StartStaleKeysReclaimer(ctx context.Context) {
+func Init() {
+	outerLocalIP := node.GetInternalIPv4Router()
+	wildcardIP := net.ParseIP("0.0.0.0")
+	ipSecReplaceStateIn(outerLocalIP, wildcardIP, false)
```
This seems fine iff this will only _add_ state for the target IP and not _replace_, which I guess is implied by the target use case here where the target IP will be different? If for some reason the keys on the filesystem change while Cilium is down and this IP is *not* new, then I suspect this may replace existing state and possibly break existing connections...?

jrfastab replied:
It can replace but only if the tuple matches, but tuple in xfrm context also includes SPI so if the cilium_host IP didn't change and keys rolled while we are off restarting we still wont replace because a key roll increments SPI.

ti-mo commented on daemon/cmd/daemon.go:
```diff
@@ -994,15 +1003,6 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	}
 
-	// Must be done at least after initializing BPF LB-related maps
-	// (lbmap.Init()).
```
Shouldn't this comment be preserved in addition to the new one you added? This context is useful to have if it's still relevant.

jrfastab committed:
cilium: Ensure xfrm state is initialized for route IP before publish

When rolling cilium-agent or doing an upgrade while running stress test
with encryption a small number of NoStateIn errors are seen. To capture
the error state (a cilium_host IP without an xfrm state rule) you need
to get into the pod near pod init and get somewhat lucky that init
took some longer time. For example I ran `ip x s` in a pod about
15seconds after launch and captured a case with new XfrmInNoErrors,
a cilium_host ip assigned, but no xfrm state rule for it. The packets
received are dropped.

The conclusion is remote nodes learn the new router IP before we have
the xfrm state rule loaded. The remote nodes then start using that
IP for the IPSec tunnel outer IP resulting in the errors when they
reach the local node without the xfrm rule yet. The errors eventually
resolve, but some packets are lost in the meantime.

The reason this happens is because first we configure the datapath
after we push node object updates. This is wrong because we need
to init the ipsec code path before we teach remote nodes about the
new IP. And second the configuration of the datapath does a lookup
in the node objects IPAddresses{} this is only populated from the
k8s watcher in the tunnel case. So we only have the fully populated
node object after we receive it through the k8s watcher. Again its
possible other nodes already have seen the event and started pushing
traffic with the new IPs.

To resolve push IPSec init code to create xfrm rules needed with
the new IPs before we publish them to the k8s node object. And
instead of pulling the IPs out of the node object simply pull them
directly from the node module. This resolves the XfrmInNoState and
XfrmIn*Policy* errors I've seen.

To reproduce the errors I can consistently reproduce with about
30 nodes, with httpperf test running from a pod in all nodes, and
then doing a 'rollout' of the cilium agent for awhile. Seems
a 2-3 hours almost ensures errors pop up. Usually the errors
happen much sooner. Initially I saw these errors on upgrade tests
which is another method to reproduce.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>
